### DISCUSSION TITLE
PDAG cseparation

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#14] [Q&A] [answered]
I am trying the `Cseparation` in PDAG object, would you please explain why I would the following error? Is it because I specified the NodeSet incorrectly?

```
import pyAgrum as gum
p = gum.PDAG()
p.addNodes(10)
p.addArc(0,1)
p.addArc(1,2)
p.cSeparation(set([0]), set([2]), set([1]))
```
<img width="1091" alt="Screenshot 2023-05-07 at 7 14 39 AM" src="https://user-images.githubusercontent.com/24616330/236674188-d6a6f29e-7795-4c92-a160-ff016e0d497b.png">

**Comment by kenneth-lee-ch** [comment]
I tried both function `Dseparation` and `Cseparation` as follows based on my guess that they both take the same arguments. The one with d-separation went through, but not for c-separation.

<img width="1102" alt="Screenshot 2023-05-07 at 3 07 32 PM" src="https://user-images.githubusercontent.com/24616330/236697733-1a6b13fa-6f6d-4cb3-8bab-e498c57d6c22.png">

**Comment by phwuil** [comment]
You are right, thanks ! `PDAG.cSeparation` was not properly wrapped. It will be in the next release !

![image](https://user-images.githubusercontent.com/5980636/237026024-a456fce6-eeeb-4fd5-b373-c1c7b18774aa.png)

**Reply by kenneth-lee-ch** [reply]
@phwuil  Thanks for looking into this. Should I just install pyagrum-nightly to get the latest version?

**Reply by phwuil** [reply]
Tag 1.8.0 is close. :-)

So tomorrow (in more than 12 hours), you should have either 1.8.0 or 1.7.1.9-nigthly that contains this commit.

**Comment by kenneth-lee-ch** [comment]
@phwuil I tried the following. Would you mind explaining why the output is true while having an edge between 0 and 1? Is there a way to check for separation for each DAG that is represented by the PDAG assuming two directions "->" and "<-" are encoded in every undirected edge ?

```
import pyAgrum as gum

p = gum.PDAG()
p.addNodes(10)
p.addEdge(0,1)
p.cSeparation([0], [1], [2])
```
<img width="1101" alt="Screenshot 2023-05-17 at 2 21 29 AM" src="https://github.com/agrumery/aGrUM/assets/24616330/b36b6c3d-acb3-4cf6-91fc-ed66b653157b">

**Comment by phwuil** [comment]
Hi @kenneth-lee-ch , believe it or not, we did not try cSeparation on such simple graphs and indeed there was a bug.  It's a bit awkward.
It is fixed and should be in nightly in less than 12 hours.

![image](https://github.com/agrumery/aGrUM/assets/5980636/8cba50fb-ee07-4065-8522-a774c56cdf6e)

**Reply by kenneth-lee-ch** [reply]
Thanks @phwuil. I updated with the latest version but I've found the following, how is `2` separated from `3`?
```
import pyAgrum as gum

p = gum.PDAG()
p.addNodes(4)
p.addEdge(0,1)
p.addArc(1,2)
p.addArc(3,2)
p.cSeparation(2, 3, [])
```
<img width="1106" alt="Screenshot 2023-05-19 at 4 52 25 AM" src="https://github.com/agrumery/aGrUM/assets/24616330/0581d1e5-09df-453a-9e5b-d0e6514b8d0f">

One more example below, how can you possibly conclude 0 is separated from 2 in this case if I have an edge between 0 and 1? I thought it should be applying d-separation in every possible DAG representation from the PDAG that has no cycles.
```
import pyAgrum as gum

p = gum.PDAG()
p.addNodes(4)
p.addEdge(0,1)
p.addArc(1,2)
p.addArc(3,2)
p.cSeparation(0, 2, [])
```
<img width="1091" alt="Screenshot 2023-05-19 at 4 57 08 AM" src="https://github.com/agrumery/aGrUM/assets/24616330/04aa1b86-39df-4264-a7d6-e6db1dc41c80">

**Reply by phwuil** [reply]
We definitely didn't do the right test suite for PDAG::cSeparation (there are about 30 tests for PDAG though ...) :-(

Thanks for taking the role of beta-testers!
(fixed in the next nightly)

![image](https://github.com/agrumery/aGrUM/assets/5980636/6acea31a-6bbe-4724-b0da-f036ec80cb6b)

**Comment by phwuil** [comment, marked as answer]
Bugfixes are included in 1.8.1
